File: BOJ/BOJ_1744_수묶기.cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<int> positive, negative;
int answer, zero;

void init() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
}

bool cmp(int &a, int &b) {
    return a > b;
}

int main(int argc, char const *argv[]) {
    init();
    cin >> N;
    for (int i = 0; i < N; i++) {
        int n; cin >> n;
        if (n > 0) {
            positive.push_back(n);
        }
        else if (n < 0) {
            negative.push_back(n);
        }
        else {
            zero++;
        }
    }
    sort(positive.begin(), positive.end(), cmp);
    sort(negative.begin(), negative.end());
    if (zero > 0 && negative.size() % 2 == 1) {
        negative.pop_back();
    } 
    for (int i = 0; i < positive.size(); i++) {
        if (positive[i] == 1) {
            answer += 1;
        }
        else if (positive[i+1] == 1) {
            answer += positive[i] + positive[i+1];
            i++;
        }
        else {
            answer += positive[i] * positive[i+1];
            i++;
        }
    }
    if (positive.size() % 2 == 1 && positive.back() != 1) {
        answer += positive.back();
    }
    for (int i = 0; i < negative.size(); i += 2) {
        if (i+1 >= negative.size()) {
            break;
        }
        answer += negative[i] * negative[i+1];
    }
    if (negative.size() % 2 == 1) {
        answer += negative.back();
    }
    cout << answer << '\n';
    return 0;
}

File: BOJ/BOJ_1744_수묶기.py
N = int(input())
nums = [int(input()) for _ in range(N)]
positive = [num for num in nums if num > 0]
negative = [num for num in nums if num < 0]
zero = nums.count(0)
answer = 0
positive.sort(reverse=True)
negative.sort()

if len(positive) > 0:
    i = 0
    while i < len(positive):
        if positive[i] == 1:
            answer += 1
        elif i+1 < len(positive) and positive[i+1] == 1:
            answer += positive[i] + positive[i+1]
            i += 1
        elif i+1 < len(positive) and positive[i+1] != 1:
            answer += positive[i] * positive[i+1]
            i += 1
        i += 1

    if len(positive) % 2 == 1 and positive[-1] != 1:
        answer += positive[-1]

if len(negative) > 0:
    if zero > 0 and len(negative) % 2 == 1:
        negative.pop()

    j = 0
    while j < len(negative):
        if j+1 >= len(negative):
            break
        answer += negative[j] * negative[j+1]
        j += 2

    if len(negative) % 2 == 1:
        answer += negative[-1]

print(answer)